clients/src/main/java/org/apache/kafka/common/utils/Checksums.java

```diff
@@ -40,11 +55,41 @@ public static void update(Checksum checksum, ByteBuffer buffer, int length) {
     public static void update(Checksum checksum, ByteBuffer buffer, int offset, int length) {
         if (buffer.hasArray()) {
             checksum.update(buffer.array(), buffer.position() + buffer.arrayOffset() + offset, length);
-        } else {
-            int start = buffer.position() + offset;
-            for (int i = start; i < start + length; i++)
-                checksum.update(buffer.get(i));
+            return;
+        }
+        if (BYTE_BUFFER_UPDATE != null && buffer.isDirect()) {
+            final int oldPosition = buffer.position();
```

Comment:
alternatively, use buffer.mark() inside try and buffer.reset() in the finally.

Reply:
We are moving `limit` as well, hence I don't see any advantage re using `mark/reset`.
Let me know If I've missed something